TLS application data isn't possible (#3043)
quicwg/base-drafts: draft-ietf-quic-tls.md

acquires handshake bytes before sending its first packet.  A QUIC server starts
the process by providing TLS with the client's handshake bytes.

At any time, the TLS stack at an endpoint will have a current sending encryption
level and receiving encryption level. Each encryption level is associated with a
different flow of bytes, which is reliably transmitted to the peer in CRYPTO
frames. When TLS provides handshake bytes to be sent, they are appended to the
current flow and any packet that includes the CRYPTO frame is protected using
keys from the corresponding encryption level.

QUIC takes the unprotected content of TLS handshake records as the content of
CRYPTO frames. TLS record protection is not used by QUIC. QUIC assembles
CRYPTO frames into QUIC packets, which are protected using QUIC packet
protection.

QUIC is only capable of conveying TLS handshake records in CRYPTO frames.  TLS
alerts are turned into QUIC CONNECTION_CLOSE error codes; see {{tls-errors}}.
TLS application data and other message types cannot be carried by QUIC at any

> encryption level and can be treated as an error if they are received.

Reviewer: How would it look like if they were received?
Wasn't @martinduke's point that QUIC doesn't even provide a way to send them at all, since there are no record headers?

Author: Sorry, I meant to say "...by a QUIC implementation".

Reviewer: Still not sure how this could happen (I don't think it could in quic-go), but I guess that depends very much on the details of the coupling between your QUIC implementation and your TLS stack.

acquires handshake bytes before sending its first packet.  A QUIC server starts
the process by providing TLS with the client's handshake bytes.

At any time, the TLS stack at an endpoint will have a current sending encryption
level and receiving encryption level. Each encryption level is associated with a
different flow of bytes, which is reliably transmitted to the peer in CRYPTO
frames. When TLS provides handshake bytes to be sent, they are appended to the
current flow and any packet that includes the CRYPTO frame is protected using
keys from the corresponding encryption level.

QUIC takes the unprotected content of TLS handshake records as the content of
CRYPTO frames. TLS record protection is not used by QUIC. QUIC assembles
CRYPTO frames into QUIC packets, which are protected using QUIC packet
protection.

QUIC is only capable of conveying TLS handshake records in CRYPTO frames.  TLS
alerts are turned into QUIC CONNECTION_CLOSE error codes; see {{tls-errors}}.

> TLS application data and other message types cannot be carried by QUIC at any

Reviewer 2: ```suggestion
TLS application data and other message types cannot be carried by QUIC
```

> encryption level and can be treated as an error if they are received from the

Reviewer 2: ```suggestion
MAY be treated as an error if they are received from the
```

Reviewer 2: Which error type? Does this result in a CONNECTION_CLOSE? Or just drop the connection? This is a local error, and therefore does not need to be communicated to the peer, so I would suggest just drop the connection. But that should be explicitly stated.

Reviewer 3: This is a clarification of fact, that a TLS stack can never return application data, unless there is a bug in the TLS stack or in the way the QUIC stack uses the TLS stack.

Therefore I think that "can be treated as an error" is fine. Changing it to "is an error" would be fine too.

Reviewer 2: @kazuho: Makes sense, but I think saying "is an error" is overloaded, since we have many types of errors. But, this isn't a big deal; I'm happy with @martinthomson's editorial decision here.

Author: I'll take Kazuho's "is an error" to avoid triggering the usual "treat X as an error of type" thinking that goes along with generating CONNECTION_CLOSE.  This is just wrong.

To Marten's point about how stacks can do this - I don't think that we generate any application_data records in our stack, but we certainly have assertions in place in case we get some.  If someone accidentally writes to the TLS socket, then we'll at catch that at the earliest possible moment.
